### ISSUE TITLE
Set up restic for offsite backups

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#821]
I've been reviewing backup solutions for a few days now, and restic has really caught my eye. Some of the nice things it does includes:

 - Incremental backups
 - Encrypted backups (with sign off from Filippo the encryption guy)
 - Mountable backups via fuse
 - Simple binary. Installation is just a wget.
 - deduplication across files and servers
 - ability to receive data from a pipe
 - works with backblaze, which appears to be the cheapest option
 - snapshots are easy to prune once old
 - active community
 - good docs

All in all, it looks pretty great. 

So, setting it up involves probably a few things:

 - [x] Create a new restic user using [the instructions in the docs](https://restic.readthedocs.io/en/latest/080_examples.html#backing-up-your-system-without-running-restic-as-root)

 - [x] Get the binary for that user and add it to the user's path.

 - [x] Establish an account at backblaze and set it up with a credit card

 - [x] Set up the backups (uses an `init` command, I think)

 - [x] Set up cronjobs for daily/weekly backups

 - [x] Set up cronjob for database dumps (weekly, probably, these are in the low hundreds of GB in size, probably)

 - [x] Kick it off.

 - [ ] Add a check command from time to time to ensure backup integrity

 - [x] Use nice and ionice (or whatever it's called) to make it low priority

 - [x] Ensure that we don't do the double file scan thing in the script. (There's a flag for this to dig up.)

**Comment by mlissner** [comment]
The `--quiet` flag prevents multiple scans of the directories: https://github.com/restic/restic/pull/1676

**Comment by mlissner** [comment]
I'm tweaking our nice and ionice params to follow the guidance in the FAQ:

    nice -n 19 ionice -c2 -n0

https://restic.readthedocs.io/en/latest/faq.html#how-to-prioritize-restic-s-io-and-cpu-time

**Comment by mlissner** [comment]
I added an extra monthly backup. Might have been useful at times. Final pruning command is:

```
 30   nice -n 19 ionice -c2 /home/restic/bin/restic forget \
 31     --prune \
 32     --keep-daily 7 \
 33     --keep-weekly 4 \
 34     --keep-monthly 2
```

**Comment by mlissner** [comment]
I've started a mega backup in a screen session. If it goes well, I'll test the prune command and then the bash script I've been working on:

#!/bin/bash

# A fairly simple backup script for restic.
# 
# For details on locking, see: https://stackoverflow.com/a/169969/64911

(
  # Wait for lock on /var/lock/.restic-backup.exclusivelock (fd 200) for 10 seconds
  
  flock -x -w 10 200 || exit 1

  # Do stuff
  export B2_ACCOUNT_ID=
  export B2_ACCOUNT_KEY=
  export RESTIC_REPOSITORY=
  export RESTIC_PASSWORD=
  nice -n 19 ionice -c2 /home/restic/bin/restic backup / \
    --exclude-caches \
    --one-file-system \
    --exclude lost+found \
    --exclude /dev \
    --exclude /home/restic \
    --exclude /home/rsnapshot \
    --exclude /media \
    --exclude /proc \
    --exclude /sata/bulk-data \
    --exclude /sys \
    --exclude /tmp \
    --quiet
  nice -n 19 ionice -c2 /home/restic/bin/restic forget \
    --prune \
    --keep-daily 7 \
    --keep-weekly 4 \
    --keep-monthly 2

) 200 > /var/lock/.restic-backup.exclusivelock

**Comment by stevesbrain** [comment]
For what it's worth, I cam here via restic/restic#1680 - just chipping in my $0.02. Pruning being slow is subjective - but my experience is that pruning a 400GB archive over the Internet (France to B2), based on a retention schedule clipping out about 70GB of data takes about 6 hours all up. Obviously many factors can affect performance, but that's approximately what I get :)

**Comment by mlissner** [comment]
Thanks for chiming in, @stevesbrain. That's...pretty bad. Seems like a lot of the performance is pretty unacceptable for a system with as much data as we have and we'll probably have to find another solution.

**Comment by stevesbrain** [comment]
@mlissner I cannot confirm how much of that is to do with B2, vs. a bare metal server (i.e. over rest-server, SFTP, etc). I know there was a discussion (in the forums, I believe) regarding the fact that restic was initially designed for more "bare metal" like that, and it wasn't until late 2017 that cloud support was introduced (or, it may have been specifically B2 cloud it was referring to). @fd0 has stated that there are improvements planned for B2 in the future, but again, no idea how far off that is, nor whether that will improve this particular use-case or not :) Either way, I hope you find the solution that works for you!

**Comment by mlissner** [comment]
Note that the command used by mlocate (which maintains the updatedb DB) is:
```
if [ -x /usr/bin/ionice ] &&
    /usr/bin/ionice -c3 true 2>/dev/null; then
    IONICE="/usr/bin/ionice -c3"
fi

flock --nonblock /run/mlocate.daily.lock $IONICE /usr/bin/updatedb.mlocate
```

So it uses `ionice -c3`, which is what we used to use, and the restic docs recommend `ionice -c2`. Hm.

**Comment by mlissner** [comment]
Thanks @fd0 for chiming in (and of course, for all your hard work)! 

I think the real issue now is incrementals. If pruning is slow, fine, we prune once a month or something. That's not the end of the world. But if incrementals take longer than the incremental period (say, one per day), then our only choice is to use a different system. That said, I'm struggling to find something that can use cloud storage in any reasonable way.

**Comment by fd0** [comment]
> But if incrementals take longer than the incremental period (say, one per day)

I'd be **very** surprised if that's the case, at least with a recent version of restic. Please let me know should you experience such a thing, it's either a corner case I haven't considered, or a bug. :)

**Comment by mlissner** [comment]
Yeah, this is with 0.8.3. It was taking days to do an incremental both with and without the `--quiet` flag. I'd love to get past this if it's possible. It was also taking about 10GB of memory. That's...fine if it's temporary for an hour or two each night, but definitely not if it's ongoing forever.

**Comment by fd0** [comment]
Ah, I remember, a **lot** of files. What's the underlying storage medium? SSD or hard  disc?

Part of the problem may be the file system, which usually isn't optimized for a very large number of files in a single directory...

**Comment by mlissner** [comment]
It's a spinning disk, and yeah, it's a LOT of files, 5-10 million, I'd say. A lot of those are in a single directory because of a dumb design I let slide a while back. Does that totally hose things?

**Comment by fd0** [comment]
Not necessarily, but it'll slow all things down. For a baseline for an initial backup you can get the timings for `tar`. For a non-initial backup, restic basically runs `stat` for each file, which will probably be slow on the file system level for such a huge directory.

The current version of restic uses way too much concurrency when archiving files (which is bad for a spinning disc, but doesn't matter on an SSD). You can try the new archiver code in branch `new-archiver`, which will read two files in parallel at most.

If you do, please report back!

**Comment by fd0** [comment]
The archiver is safe (it won't eat your data), but I though you give it a try on a test bucket (or local directory) or so. Just to get a feeling if it'd improve things.

There aren't any binaries yet, we only build beta binaries for the master branch. So once it's merged, you'll get a binary :)

**Comment by stevesbrain** [comment]
@mlissner But the regular build commands work fine for the new-archiver branch, just to be clear :) Very very quick process!

**Comment by mlissner** [comment]
Just FYI, this is on hold until a disk-intensive process finishes and we rebuild a RAID array. No point in doing performance tests with all this other stuff going on.

**Comment by mlissner** [comment]
OK, back on this project. Our RAID still isn't rebuilt properly, but I've decided that's all the more reason to do this project (sigh). 

Got the latest version of restic with:

    cd /home/restic/bin
    wget https://github.com/restic/restic/releases/download/v0.9.1/restic_0.9.1_linux_386.bz2
    sudo bzip2 -d restic_0.9.1*

Now have it running with:

    sudo ./restic-backup.sh

**Comment by mlissner** [comment]
I'm checking progress with:

```
rpid=$(ps ax -o fname,pid | sed -e '/restic /!d;s/.* //')
lsof -p $rpid -n | grep REG | sed -e '/tmp.restic-/d'
```

**Comment by mlissner** [comment]
OK, it looks like the new archiver code in the latest restic will be good enough for us (we also did some architectural changes to have plenty of RAM laying around). We did an incremental backup last night that worked without a hitch. 

I'm currently backing up the DB using the `--stdin` flag. Left to do is:

- [ ] Test a restore
- [ ] Cronify the DB backups (note that you can't do `ALTER` commands during an active `pg_dump` command, so these will need to happen when we're not updating the DB.

**Comment by mlissner** [comment]
I'm experimenting with the `forget --prune` incantation because a couple of aborted backups wound up in the snapshot list. Looks like performance is going to be pretty bad with these. Our current one is looking like it will take about 19 hours to complete. 

I asked on their forum to see if there is advice: 

https://forum.restic.net/t/pruning-multiple-snapshots-simultaneously-fasters/837

Perhaps it can be resolved by doing less frequent prunes or by separating our changing data (the DB, mostly) from our incremental data (the PDFs, mostly).

**Comment by mlissner** [comment]
OK, so the prune took *days*, and the forum post above indicates that they see this as a big priority. So for now, since we don't delete much anyway, I think the best we can do is just not prune and pay for the extra data. I commented out the pruning command from our snapshot script.

This means that there's two things left to do. One is test a restore and the other is to update our TOS. @anseljh or @brianwc, I wonder if you guys have experience with/advice for this. We're now uploading client-side-encrypted data to backblaze for our backups. Does this need to go into our TOS? Should it? We're technically sharing data with them, but for all backblaze knows, it's just 4.5TB of encrypted blobs.  

As for the restore, I just tested a few things and I'm pretty confident we're good here. Calling that done too, though in a perfect world we'd do a *FULL* restore.

So our backups are done with the exception of a possible TOS update. Ansel, Brian, any opinions?

**Comment by mlissner** [comment]
OK, so the prune took *days*, and the forum post above indicates that they see this as a big priority. So for now, since we don't delete much anyway, I think the best we can do is just not prune and pay for the extra data. I commented out the pruning command from our snapshot script.

This means that there's two things left to do. One is test a restore and the other is to update our TOS. @anseljh or @brianwc, I wonder if you guys have experience with/advice for this. We're now uploading client-side-encrypted data to backblaze for our backups. Does this need to go into our TOS? Should it? We're technically sharing data with them, but for all backblaze knows, it's just 4.5TB of encrypted blobs.  

As for the restore, I just tested a few things and I'm pretty confident we're good here. Calling that done too, though in a perfect world we'd do a *FULL* restore.

So our backups are done with the exception of a possible TOS update. Ansel, Brian, any opinions?

**Comment by mlissner** [comment]
I went ahead and did the TOS update: https://www.courtlistener.com/terms/#privacy because why not. 

Marking this done. Thanks to the other folks that chimed in here to make this possible, especially to @fd0 for the amazing restic backup utility. I've opened a new bug https://github.com/freelawproject/courtlistener/issues/856 to take care of pruning once it's fixed in restic.
